### ISSUE TITLE
Check implementation validity for an overloaded function

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ty#3118]
This is an idea I thought of after reading [this comment](https://github.com/astral-sh/ty/issues/3086#issuecomment-4099514234), and I figured I might as well throw it out there since it could be cool, even if it probably won't happen.

This is different and more advanced version of #109, where instead of just making sure the signature works, ty would also make sure the body of the function works.

In pseudocode, the process would be:
```
If a function has overloads and an implementation
    For every overload definition
        Type check the function as if it had the overload definition and collect all possible return types
        If any return type is not assignable to the overload return type, raise a diagnostic
```

Because the implementation usually combines all of the overloads with unions, as long as one overload is satisfied by the implementation it will pass. This check would allow detecting those functions that partially/incorrectly implement their overloads, as in https://github.com/JelleZijlstra/unsoundness/blob/main/examples/overload/incorrect_impl.py .

There are several downsides to this though:
- I'm not very familiar with ty's internals, so implementing this could range anywhere from trivial to large internal rewrites.
- Very heavily relies on unreachable detection, so could be very prone to false positives if ty does not see branches as unreachable, ie due to runtime only things.
- Would multiply the check time of functions by the number of overloads they have, since the function would have to be re-checked in full per overload, which would be very expensive for complex functions.

Overall the downsides probably outweigh the benefits, but I figured I'd throw this out there anyways.

Some examples of this:
```py
from typing import overload

@overload
def correct(x: int) -> int: ...
@overload
def correct(x: str) -> str: ...
def correct(x: int | str) -> int | str:  # No error, function implements overloads correctly
    return x

# Checking overload
def correct(x: int) -> int:
    return x  # `int` is assignable to `int`, no error

# Checking overload
def correct(x: str) -> str:
    return x  # `str` is assignable to `str`, no error
```

```py
from typing import overload

@overload
def wrong(x: int) -> str: ...
@overload
def wrong(x: str) -> int: ...
def wrong(x: int | str) -> int | str:  # Error, overload `(str) -> int` can return `str` instead of `int`
    if isinstance(x, int):
        return str(x)
    else:
        return x

# Checking overload
def wrong(x: int) -> str:
    if isinstance(x, int):
        return str(x)  # `str` is assignable to `str`, no error
    else:
        return x  # Unreachable, so `Never`, which is assignable to `str`, no error

# Checking overload
def wrong(x: str) -> int:
    if isinstance(x, int):
        return str(x)  # Unreachable, so `Never`, which is assignable to `int`, no error
    else:
        return x  # `str` which is not assignable to `int`, so error
```

**Comment by dhruvmanila** [comment]
Thank you for providing the details, I think this is a duplicate of https://github.com/astral-sh/ty/issues/109.

**Comment by AlexWaygood** [comment]
@dhruvmanila I don't think this is a duplicate. #109 is about comparing the implementation signature with the overload signatures to check for inconsistencies: that's mandated by the spec, and something all other type checkers do. This issue proposes something much more advanced: it proposes verifying the function body of the implementation against each of the signatures, which would be much more expensive but also much more exhaustive. This is not mandated by the spec and not done by any existing type checker to my knowledge, but it would certainly be a cool feature.

**Comment by dhruvmanila** [comment]
(Oops, sorry, somehow I completely missed that part)
